lib/src/rules/iterable_contains_unrelated_type.dart

```diff
@@ -0,0 +1,164 @@
+
+**BAD:**
+```
+void someFunction() {
+  var list = <int>[];
+  if (list.contains('1')) print('someFunction'); // LINT
+}
+```
+
+**BAD:**
+```
+void someFunction3() {
+  List<int> list = <int>[];
+  if (list.contains('1')) print('someFunction3'); // LINT
+}
+```
+
+**BAD:**
+```
+void someFunction8() {
+  List<DerivedClass2> list = <DerivedClass2>[];
+  DerivedClass3 instance;
+  if (list.contains(instance)) print('someFunction8'); // LINT
+}
+```
+
+**GOOD:**
+```
+void someFunction10() {
+  var list = [];
+  if (list.contains(1)) print('someFunction10'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction1() {
+  var list = <int>[];
+  if (list.contains(1)) print('someFunction1'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction4() {
+  List<int> list = <int>[];
+  if (list.contains(1)) print('someFunction4'); // OK
+}
+```
+
+**GOOD:**
+```
+void someFunction5() {
+  List<ClassBase> list = <ClassBase>[];
+  DerivedClass1 instance;
+  if (list.contains(instance)) print('someFunction5'); // OK
+}
+
+abstract class ClassBase {}
+
+class DerivedClass1 extends ClassBase {}
+```
+
+**GOOD:**
+```
+void someFunction6() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass2 instance;
+  if (list.contains(instance)) print('someFunction6'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass2 extends ClassBase with Mixin {}
+```
+
+**GOOD:**
+```
+void someFunction7() {
+  List<Mixin> list = <Mixin>[];
+  DerivedClass3 instance;
+  if (list.contains(instance)) print('someFunction7'); // OK
+}
+
+abstract class ClassBase {}
+
+abstract class Mixin {}
+
+class DerivedClass3 extends ClassBase implements Mixin {}
+```
+''';
+
+class IterableContainsUnrelatedType extends LintRule {
+  _Visitor _visitor;
+
+  IterableContainsUnrelatedType() : super(
+          name: 'iterable_contains_unrelated_type',
+            description: _desc,
+            details: _details,
+      group: Group.errors,
+      maturity: Maturity.experimental) {
+    _visitor = new _Visitor(this);
+  }
+
+  @override
+  AstVisitor getVisitor() => _visitor;
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (!_isParameterizedIterableContainsInvocation(node)) {
+      return;
+    }
+
+    ParameterizedType type = node.target.bestType;
+    Expression argument = node.argumentList.arguments.first;
+    if (unrelatedTypes(argument.bestType, type.typeArguments.first)) {
+      rule.reportLint(node);
+    }
+  }
+}
+
+bool _isParameterizedIterableContainsInvocation(MethodInvocation node) =>
+    node.target != null &&
+        node.methodName.name == 'contains' &&
+        _implementsParameterizedIterable(node.target.bestType) &&
+        node.argumentList.arguments.length == 1 &&
+        node.operator.type == TokenType.PERIOD;
+
+bool _implementsParameterizedIterable(DartType type) =>
```

Comment:
It would be good to add a test of a class that indirectly implements Iterable:

abstract class A<E> implements Iterable<E> {}
abstract class B<E> extends A<E> {}
bool takesB(B<int> b) => b.contains('a');